### PR TITLE
PostgreSQL monitoring integration

### PR DESCRIPTION
**Pull request by nmtien**

### CODE REVIEW

**Comment by nmtien** [comment]
i rename metric, update description based on your comments.

File: postgresql/db-activity-status-8-3.yml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        description: Number of backends executing a query
+
+      - name: transaction.idle.count
+        source: idle_in_transaction
+        type: gauge
+        label: Idle backend
+        description: Number of backend is in a transaction, but is not currently executing a query
+
+    tag:
+      - name: activity.database
```
Comment (sivasamyk) [inline postgresql/db-activity-status-8-3.yml]: Prefix `postgresql` (metric namespace) to all tags.

File: postgresql/db-common-bgwriter-status-9-2.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  dbDriverClass: org.postgresql.Driver
+  dbUser: ${POSTGRESQL_DB_USER}
+  dbPassword: ${POSTGRESQL_DB_PASSWORD}
+  dbAdditionalConnectionParams: ${POSTGRESQL_DB_ADDITIONAL_PARAMS}
+  #  dbVerticalModel="true" - because output of SHOW STATUS command gives each attribute in its own row
+  # (1st column being attribute name, 2nd column being attribute value)
+  dbVerticalModel: true
+
+require:
+  - className: com.sematext.spm.client.postgresql.PostgreSQLVersionCheck
```
Comment (sivasamyk) [inline postgresql/db-common-bgwriter-status-9-2.yml]: Are these 2 attributes applicable only for 9.2 or all version greater than 9.2?

File: postgresql/db-common-status-9-2.yml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        type: long_counter
+        unit: bytes
+        label: Temporary written bytes
+        description: Total amount of data written to temporary files by queries in this database. All temporary files are counted, regardless of why the temporary file was created, and regardless of the log_temp_files setting
+
+      - name: database.temp.files
+        source: temp_files
+        type: long_counter
+        label: Temporary files
+        description: Number of temporary files created by queries in this database. All temporary files are counted, regardless of why the temporary file was created (e.g., sorting or hashing), and regardless of the log_temp_files setting
```
Comment (sivasamyk) [inline postgresql/db-common-status-9-2.yml]: Check if the YAML is valid. If the description has special characters you might need to enclose inside double quotes

File: postgresql/db-common-status.yml [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        source: tup_updated
+        type: long_counter
+        label: Rows updated
+        description: Number of rows updated by queries in this database
+
+      - name: database.rows.deleted
+        source: tup_deleted
+        type: long_counter
+        label: Rows deleted
+        description: Number of rows deleted by queries in this
```
Comment (sivasamyk) [inline postgresql/db-common-status.yml]: `database`  is missing

File: postgresql/db-common-bgwriter-status-9-2.yml (hunk 2 above)
Comment (nmtien) [reply postgresql/db-common-bgwriter-status-9-2.yml]: Thanks for good question, i check and correct all version range check. This is for all version greater or equals to 9.2

File: postgresql/db-activity-status-8-3.yml (hunk 1 above)
Comment (nmtien) [reply postgresql/db-activity-status-8-3.yml]: @bsmid should we add all metrics with this prefix?

File: postgresql/db-common-status-9-2.yml (hunk 3 above)
Comment (nmtien) [reply postgresql/db-common-status-9-2.yml]: I quote it, i also quote multiple line query and check all YAML file with https://codebeautify.org/yaml-validator

File: postgresql/db-common-status.yml (hunk 4 above)
Comment (nmtien) [reply postgresql/db-common-status.yml]: add it, thanks.

File: postgresql/db-activity-status-8-3.yml (hunk 1 above)
Comment (bsmid) [reply postgresql/db-activity-status-8-3.yml]: Metrics shouldn't get this prefix; as Siva said - tags should get it. Also `activity` part isn't needed, tag name should probably be `postgresql.database`.

File: postgresql/db-common-bgwriter-status.yml [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        source: buffers_checkpoint
+        type: long_counter
+        label: Checkpoint buffers
+        description: Number of buffers written during checkpoints
+
+      - name: bgwriter.buffers.clean
+        source: buffers_clean
+        type: long_counter
+        label: Clean buffers
+        description: Number of buffers written by the background writer
```
Comment (bsmid) [inline postgresql/db-common-bgwriter-status.yml]: Sounds like possibly description of some other metric? Also, should `clean` part use past tense like in other metrics (so it is `...cleaned`?

File: postgresql/db-common-bgwriter-status.yml [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+      - name: bgwriter.buffers.maxwritten.clean
```
Comment (bsmid) [inline postgresql/db-common-bgwriter-status.yml]: If it is about stopped cleanings, maybe use a name like `bqwriter.bufferes.cleaning.maxwritten.stopped`.

File: postgresql/db-common-status.yml [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+observation:
+  - name: database
+    metricNamespace: postgresql
+
+    metric:
+      - name: database.connections.active.count
+        source: numbackends
+        type: gauge
+        label: Active connections
+        description: Number of backends currently connected to this database
```
Comment (bsmid) [inline postgresql/db-common-status.yml]: Is it "number of backends" or "number of connections"?

File: postgresql/db-common-status.yml (hunk 4 above)
Comment (bsmid) [inline postgresql/db-common-status.yml]: "in this..." database missing at the end?

File: postgresql/db-connection-status.yml [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  #  dbVerticalModel="true" - because output of SHOW STATUS command gives each attribute in its own row
+  # (1st column being attribute name, 2nd column being attribute value)
+  dbVerticalModel: true
+
+observation:
+  - name: connection
+    metricNamespace: postgresql
+
+    metric:
+      - name: connections.max
```
Comment (bsmid) [inline postgresql/db-connection-status.yml]: Here metric is named `connections.max`, while file `db-common-status.yml` has metrics like `database.connections.active.count` - note the `database` prefix. Are metrics from these two configs on different level? If metrics in `db-common-status.yml` are about specific database instance, are they missing some `database` tag?

File: postgresql/db-activity-status-10.yml [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+      - name: transaction.idle.count
+        source: idle_in_transaction
+        type: gauge
+        label: Idle backend
+        description: Number of backend is in a transaction, but is not currently executing a query
+
+    tag:
+      - name: activity.database
+        value: eval:${database}
```
Comment (bsmid) [inline postgresql/db-activity-status-10.yml]: Does this produce anything? There should be metric named `database` which would provide value here, I don't see it? And looking into other config files, I see the same problem. But maybe I am missing something?

File: postgresql/db-index-status.yml [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+  dbUser: ${POSTGRESQL_DB_USER}
+  dbPassword: ${POSTGRESQL_DB_PASSWORD}
+  dbAdditionalConnectionParams: ${POSTGRESQL_DB_ADDITIONAL_PARAMS}
+
+observation:
+  - name: indexes
+    metricNamespace: postgresql
+
+    metric:
+      - name: indexes.all.table
```
Comment (bsmid) [inline postgresql/db-index-status.yml]: What `all` part means in this case? Perhaps it can be omitted?

File: postgresql/db-count-status.yml [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+        send: false
+        label: Schema name
+
+      - name: table.count
+        source: table_count
+        type: gauge
+        label: Table count
+
+    tag:
+      - name: count.schema
```
Comment (bsmid) [inline postgresql/db-count-status.yml]: This should probably be `postgresql.schema`. Part `count` is useless here.

File: postgresql/db-function-status.yml [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+      - name: function.time.self
+        source: self_time
+        type: counter
+        unit: ms
+        label: function self time
+        description: Total time spent in this function itself, not including other functions called by it, in milliseconds
+
+    tag:
+      - name: function.schema
```
Comment (bsmid) [inline postgresql/db-function-status.yml]: Should probably be just `postgresql.schema`.

File: postgresql/db-function-status.yml [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+        value: eval:${schema}
+      - name: function.name
```
Comment (bsmid) [inline postgresql/db-function-status.yml]: Should be `postgresql.function`.

File: postgresql/db-index-status.yml [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+        description: Number of index entries returned by scans on this index
+
+      - name: indexes.all.rows.feched
+        source: idx_tup_fetch
+        type: long_counter
+        label: Fetched rows
+        description: Number of live rows fetched by index scans
+
+    tag:
+      - name: indexes.all.table
```
Comment (bsmid) [inline postgresql/db-index-status.yml]: I guess this is just a plain "table" tag, not some special "index all table" that has no relation to DB tables. If yes, name should be `postgresql.table`.

File: postgresql/db-index-status.yml [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+        value: eval:${table}
+      - name: indexes.all.schema
```
Comment (bsmid) [inline postgresql/db-index-status.yml]: Just `postgresql.schema`?

File: postgresql/db-index-status.yml [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+        value: eval:${schema}
+      - name: indexes.all.index
```
Comment (bsmid) [inline postgresql/db-index-status.yml]: Just `postgresql.index`? Btw, I will not repeat similar suggestions in other places, you get the idea.

File: postgresql/db-rel-status.yml [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+  dbUser: ${POSTGRESQL_DB_USER}
+  dbPassword: ${POSTGRESQL_DB_PASSWORD}
+  dbAdditionalConnectionParams: ${POSTGRESQL_DB_ADDITIONAL_PARAMS}
+
+observation:
+  - name: usertables
+    metricNamespace: postgresql
+
+    metric:
+      - name: tables.user.table
```
Comment (bsmid) [inline postgresql/db-rel-status.yml]: What is the point of `user` part? Could we omit it? Same for other metrics.

File: postgresql/db-rel-status.yml [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+        send: false
+        label: Table name
+
+      - name: tables.user.schema
+        source: schemaname
+        type: text
+        send: false
+        label: Schema name
+
+      - name: tables.user.sequential.scan
```
Comment (bsmid) [inline postgresql/db-rel-status.yml]: Typically would be other way around - `tables.scan.sequential`. And other metrics are missing `scan` part.

File: postgresql/db-size-status.yml [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+    metricNamespace: postgresql
+
+    metric:
+      - name: size.table
+        source: relname
+        type: text
+        send: false
+        label: Table name
+
+      - name: table.size
```
Comment (bsmid) [inline postgresql/db-size-status.yml]: Maybe name it `table.disk.size`.

File: postgresql/db-size-status.yml [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+        source: table_size
+        type: gauge
+        unit: bytes
+        label: Table size
+        description: Disk space used by the a table, excluding indexes (but including TOAST, free space map, and visibility map)
+
+      - name: index.size
```
Comment (bsmid) [inline postgresql/db-size-status.yml]: `index.disk.size`

File: postgresql/db-size-status.yml [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+        source: index_size
+        type: gauge
+        unit: bytes
+        label: Index size
+        description: Total disk space used by indexes attached to the a table
+
+      - name: total.size
```
Comment (bsmid) [inline postgresql/db-size-status.yml]: `table.total.disk.size`.

File: postgresql/db-size-status.yml [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+        source: total_size
+        type: gauge
+        unit: bytes
+        label: Total size
+        description: Total disk space used by the a table, including all indexes and TOAST data
+
+    tag:
+      - name: size.table
```
Comment (bsmid) [inline postgresql/db-size-status.yml]: Just `postgresql.table`.

File: postgresql/db-statio-status.yml [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+        send: false
+        label: Table name
+
+      - name: statio.schema
+        source: schemaname
+        type: text
+        send: false
+        label: Schema name
+
+      - name: statio.heap.blocks.read
```
Comment (bsmid) [inline postgresql/db-statio-status.yml]: How about `table.io.heap.blocks.read`? Same for other metrics in this file.

File: postgresql/db-statio-status.yml [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+        source: heap_blks_read
+        type: long_counter
+        label: Read disk blocks
+        description: Number of disk blocks read from a table
+
+      - name: statio.heap.blocks.git
```
Comment (bsmid) [inline postgresql/db-statio-status.yml]: `git`-> `hit` + `table.io....` as prefix.

File: postgresql/db-activity-status-8-3.yml (hunk 1 above)
Comment (nmtien) [reply postgresql/db-activity-status-8-3.yml]: Ok Thanks @sivasamyk @bsmid

File: postgresql/db-activity-status-10.yml [hunk 25]
```diff
@@ -0,0 +1,60 @@
...
+        source: datname
+        type: text
+        send: false
+        label: Database
+        description: Name of the database this backend is connected to
+
+      - name: transaction.open.count
+        source: open
+        type: long_gauge
+        label: Active backends
```
Comment (sivasamyk) [inline postgresql/db-activity-status-10.yml]: All labels shall start with lower case and use only lower case except for abbreviations.

File: postgresql/db-activity-status-10.yml [hunk 26]
```diff
@@ -0,0 +1,60 @@
...
+        description: Number of backends executing a query
+
+      - name: transaction.idle.count
+        source: idle_in_transaction
+        type: long_gauge
+        label: Idle backend
+        description: Number of backend is in a transaction, but is not currently executing a query
+
+    tag:
+      - name: posgresql.database
```
[the remark this answers is not in the document]
Comment (nmtien) [reply postgresql/db-activity-status-10.yml]: Updated. Thanks

File: postgresql/db-statio-status.yml (hunk 23 above)
Comment (nmtien) [reply postgresql/db-statio-status.yml]: Change to statio.heap.blocks.read

File: postgresql/db-common-archiver-status.yml [hunk 27]
```diff
@@ -0,0 +1,28 @@
+type: db
+data:
+  query: SELECT * FROM pg_stat_archiver;
+  dbUrl: jdbc:postgresql://${POSTGRESQL_DB_HOST_PORT}/${POSTGRESQL_DB_DATABASE}
+  dbDriverClass: org.postgresql.Driver
+  dbUser: ${POSTGRESQL_DB_USER}
+  dbPassword: ${POSTGRESQL_DB_PASSWORD}
+  dbAdditionalConnectionParams: ${POSTGRESQL_DB_ADDITIONAL_PARAMS}
+  #  dbVerticalModel="true" - because output of SHOW STATUS command gives each attribute in its own row
```
Comment (bsmid) [inline postgresql/db-common-archiver-status.yml]: This comment should be removed. It is present in other files, remove from there too please.

Comment (nmtien) [reply postgresql/db-common-archiver-status.yml]: Removed this comment on all postgresql yml files.

**Pull request merged**